[NFC] Distinguish references to names from declarations of those names by type

beccadax opened the pull request:
This is a refactoring in support of the [module qualification feature](https://forums.swift.org/t/pitch-fully-qualified-name-syntax/28482/86) I've previously pitched on Evolution.

The eventual goal for that feature is to allow any place that *uses* a name for a declaration defined elsewhere to qualify it with a module name. That is, all of these should allow modules:

```swift
print(SomeModule::x)   // The `x` visible from SomeModule
({ () -> SomeModule::T in ... })(x)   // The `T` from SomeModule
T.SomeModule::init()   // The `T.init()` declared in SomeModule
```

But things like this should not:

```swift
func SomeModule::fn() { ... }   // defining a function in a different module???
let SomeModule::x = 1    // defining a constant in another module???
```

To support this distinction, this PR adds a `DeclNameRef` type to the compiler. This type currently just wraps `DeclName`, but in the future it will also be able to store a module name when the user module-qualifies a name. Declarations will continue to store a `DeclName` or `Identifier` for their own names, but name lookup entry points—and therefore many other AST nodes like `UnresolvedDeclRefExpr`s and `TypeRepr`s—will switch over to using `DeclNameRef`. Thus, this PR will establish the pathways through which module qualification information will flow without actually making any functional changes to the compiler. This will reduce the amount of merge conflicts during the feature's design and review.

~~While this PR is in progress, `DeclNameRef`'s constructors will be implicit but deprecated, and a `DeclNameRef_` function is used whenever we want to explicitly construct a `DeclNameRef`. Before I merge this commit, I will make the `DeclNameRef` constructors explicit, rename all `DeclNameRef_`s to `DeclNameRef`, and delete the `DeclNameRef` function.~~

beccadax committed:
[NFC] Make base-name behavior of lookup table loading explicit

We were unnecessarily converting back and forth between base names and full names, which seemed a bit silly.

brentdax commented:
There's cleanup to do on this branch, but it's getting towards the end.

brentdax commented on include/swift/AST/DiagnosticsCommon.def:
```diff
@@ -166,7 +166,7 @@ NOTE(kind_declname_declared_here,none,
         "ignoring associated type %0 in favor of module-scoped property "
         "wrapper %0; please qualify the reference with %1",
-        (DeclName, Identifier))
+        (DeclNameRef, Identifier))
```
Check this.

brentdax commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -207,7 +207,7 @@ ERROR(lex_conflict_marker_in_file,none,
 
 NOTE(note_in_decl_extension,none,
-     "in %select{declaration|extension}0 of %1", (bool, Identifier))
+     "in %select{declaration|extension}0 of %1", (bool, DeclNameRef))
```
Honestly, I’m surprised this isn’t a TypeRepr for extensions.

brentdax commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -78,33 +78,34 @@ ERROR(could_not_find_value_subscript,none,
 
 ERROR(could_not_find_enum_case,none,
-      "enum type %0 has no case %1; did you mean %2", (Type, DeclName, DeclName))
+      "enum type %0 has no case %1; did you mean %2",
```
Unrelated: why no question mark?

brentdax replied:
Filed as https://bugs.swift.org/browse/SR-11919.

brentdax commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -127,35 +128,35 @@ ERROR(unexpected_arguments_in_contextual_member,none,
       "%0 is inaccessible due to "
       "'%select{private|fileprivate|internal|%error|%error}1' protection level",
-      (DeclName, AccessLevel))
+      (DeclBaseName, AccessLevel))
```
These are all cases where we actually found a declaration. Should we use the declaration’s actual name instead of the name it was looked up with?

brentdax commented:
@swift-ci please test

brentdax commented:
With apple/llvm-project#430

@swift-ci please test

brentdax commented:
I'm going to try running tests from the llvm-project side in apple/llvm-project#430.

brentdax commented:
With apple/llvm-project#430

@swift-ci please test

brentdax commented:
Notes for reviewers:

* This is mostly boring NFC type name changes.

* In general, `DeclNameRef` should be used where the developer has written the name of a type that is declared elsewhere. However, at various points we need to take a name from a declaration and look it up as though the user wrote it, and at other points we need to strip away the distinction between names and references to names so we can work with them as equals, and there are some places in constraint solving particularly where we end up losing the original names, so you'll sometimes see conversions back and forth.

* In many of the commits, you'll see `DeclNameRef_`. These mark places where I eventually inserted an explicit conversion. There are no `DeclNameRef_`s left in the final commit, and the function they used to call is gone.

* The easiest thing to do might be to read the smaller commits in more detail and skim the large commits for parts that look non-trivial.

CodaFi reviewed (approved):
I can't spot anything wrong, but then again this is a *lot* of code.  Can the smaller refactorings be merged off separately?  Ideally this patch would just be the renaming and the type plumbing afterwards.

brentdax commented:
@CodaFi I could pull b1fbdbfe (Miscellaneous improvements to Playground and PCMacro transforms) into a separate followup PR; that part is +84 -88 and nothing else needs it. I don't know if there's anything else I can easily extract, though. Any suggestions?

brentdax commented:
Hmm...it'd be trickier, but these could probably be extracted and landed in precursors:

* Introduction of {UnresolvedDeclRefExpr,UnresolvedDotExpr}::createImplicit()
* PreCheckExpression::simplifyTypeExpr changes to use an ASTContext instead of a raw DiagnosticEngine in helpers
* EnumElement changes to use a DeclName in more places

Some of these may be difficult to separate, but I'll see what I can do.

CodaFi commented:
Those match my expectations.  They should be simple NFC commits.

beccadax committed:
[NFC] Add helpers for unresolved AST node synthesis

This change adds UnresolvedDotExpr::createImplicit() and UnresolvedDeclRefExpr::createImplicit() helpers. These calls simplify several tedious bits of code synthesis that would otherwise become even more tedious with DeclNameRef in the picture.

beccadax committed:
[NFC] Use ASTContext::Id_Void constant in simplifyTypeExpr check

beccadax committed:
[NFC] Make EnumElementPattern carry a DeclName

This has the side effect of threading compound name support through more enum pattern code, although it’s still not complete.

beccadax committed:
[NFC] Extract Parser::parseUnqualifiedDeclBaseName

beccadax committed:
[NFC] Convert TypeRepr to use DeclName(Loc)?

Replaces `ComponentIdentTypeRepr::getIdentifier()` and `getIdLoc()` with `getNameRef()` and `getNameLoc()`, which use `DeclName` and `DeclNameRef` respectively.

beccadax committed:
[NFC] Update uses in TypeCheckConstraints

Seems to have been a mismerge or something?

beccadax committed:
[NFC] Improve identifier usage in AST transforms

Change the various AST transforms to use prebuilt DeclName constants more heavily rather than an ad-hoc mix of Identifiers, string literals, std::strings, and StringRefs with questionable memory management.

beccadax committed:
[NFC] Document a weird designated types-related accessor

This feature is barely used and needs refactoring that would be outside the scope of this series of commits (see https://bugs.swift.org/browse/SR-11241).

beccadax committed:
[NFC] Add DeclNameRef type to the compiler

This type wraps a DeclName, indicating that it is a reference to a declaration that exists somewhere else and it requires slightly “fuzzy” comparison (i.e. if it’s not compound, only the base names should be compared). DeclName::matchesRef() and MemberLookupTable::find() both now take a DeclNameRef instead of a DeclName.

This commit temporarily allows implicit conversion from DeclName; I’ll flip the switch on that in a later commit.

beccadax committed:
[NFC] Add a spare bit to Identifier

So that DeclName can have a spare bit, allowing it to be used in places where we’re currently using PointerIntPair<Identifier, 1> or PointerUnion<Identifier, Whatever>.

beccadax committed:
[NFC] Remove DeclNameRef staging calls

beccadax committed:
[NFC] DeclNameRef-ify qualified and unqualified lookup

beccadax committed:
[NFC] Thread DeclNameRef through most of the compiler

This huge commit contains as many of the mechanical changes as possible.

beccadax committed:
[NFC] Take DeclNameRef in UnqualifiedLookup and lookupQualified()

beccadax committed:
[NFC] Miscellaneous improvements to Playground and PCMacro transforms

brentdax commented:
@CodaFi Okay, I think you'll find this one more manageable:

* There is one enormous commit, 150e4e "[NFC] Thread DeclNameRef through most of the compiler", which is intended to contain no interesting changes.

* There are two large but hopefully workable commits which convert important AST nodes from Identifier to DeclName (not DeclNameRef; that comes later) and SourceLoc to DeclNameLoc:
    * EnumElementPattern in a3afd8 "[NFC] Make EnumElementPattern carry a DeclName"
    * TypeRepr in 681be3 "[NFC] Convert TypeRepr to use DeclName(Loc)?"

* Everything else should be *relatively* small and focused.

brentdax commented:
With apple/llvm-project#430

@swift-ci please test

brentdax commented:
Needed yet another rebase.

With apple/llvm-project#430

@swift-ci please smoke test

brentdax commented:
With apple/llvm-project#430

@swift-ci please test compiler performance

brentdax commented:
With apple/llvm-project#430

@swift-ci please test source compatibility

brentdax commented:
With apple/llvm-project#430

@swift-ci please smoke test compiler performance
